### PR TITLE
Configure request timeouts (close #2940)

### LINKED ISSUE #2940: Allow configurable timeouts for page and XHR requests
### Are you requesting a feature or reporting a bug?
Bug

### What is the current behavior?
Any XHR that takes longer than 20 seconds will make tests depending on parsing the response fail

### What is the expected behavior?
Either the page load ltimeout or another timeout should be available to make testcafe not fail when XHR take longer than 20 seconds

### How would you reproduce the current behavior (if this is a bug)?
Run the below test which loads twp page that retrieve a simple JSON from a remote service using XHR. The first page XHR response returns in less than 1 sec. The second page XHR response returns in 25 seconds. The first succeeds but the second fails because testcafe gives up on XHR that take over 20 seconds. 
```
import { Selector } from 'testcafe';

fixture `Testcafe XHR support POC`

test('the label jsonDisplay should show up the fast XHR response', async t => {
  // The below succeeds 
  await t.navigateTo('http://www.nestorurquiza.com/xhr.html');
  const jsonDisplay = Selector('#jsonDisplay').addCustomDOMProperties({
    innerHTML: el => el.innerHTML
  });
  await t.expect(jsonDisplay.visible).ok();
  await t.expect(jsonDisplay.innerHTML).contains('hello');
});

test('the label jsonDisplay should show up the slow XHR response', async t => {
  // The below fails
  await t.navigateTo('http://www.nestorurquiza.com/slowXHR.html');
  const jsonDisplay = Selector('#jsonDisplay').addCustomDOMProperties({
    innerHTML: el => el.innerHTML
  });
  await t.expect(jsonDisplay.visible).ok();
  await t.expect(jsonDisplay.innerHTML).contains('hello');
});
```
Using the following command the test ends with the below error:
```
$ ./node_modules/.bin/testcafe --selector-timeout 60000 --assertion-timeout 60000 --page-load-timeout 60000 chrome  tests/slow-xhr.test.js
 Running tests in:
 - Chrome 69.0.3497 / Linux 0.0.0

  XHR
 ✓ the label jsonDisplay should show up the fast XHR response
 ✖ the label jsonDisplay should show up the slow XHR response

   1) AssertionError: expected 'Result of request should appear here ...' to include 'hello'

      Browser: Chrome 69.0.3497 / Linux 0.0.0

         17 |  await t.navigateTo('http://www.nestorurquiza.com/slowXHR.html');
         18 |  const jsonDisplay = Selector('#jsonDisplay').addCustomDOMProperties({
         19 |    innerHTML: el => el.innerHTML
         20 |  });
         21 |  await t.expect(jsonDisplay.visible).ok();
       > 22 |  await t.expect(jsonDisplay.innerHTML).contains('hello');
         23 |});
         24 |

         at contains (/home/nurquiza/workspace/lms-e2e/tests/slow-xhr.test.js:22:41)



 1/2 failed (1m 05s)
```
Note that the command sets to 1 minute all the available timeouts. Optionally and for the record, see below the html of this page in case you want to host it yourself. Change the mocky-delay param to have a page to render in 1ms and another in 25s (If you host it you will need to either have the XHR service in the same sub-domain or provide CORS support for the remote service):

```
<html>
<head>
<title>XHR example</title>
</head>
<body>
<label id="jsonDisplay">Result of request should appear here ...</label>
<script>
const url = 'https://www.mocky.io/v2/5185415ba171ea3a00704eed?mocky-delay=1ms';
const jsonDisplay = document.querySelector('#jsonDisplay');
fetch(url).then(function(response) {
  response.text().then(function(text) {
    jsonDisplay.innerHTML = text;
  });
});
</script>
</body>
```

#### Provide the test code and the tested page URL (if applicable)

See above. Note that this is an issue that will result in  tests suddenly not passing after a release just because the responses from the database are slower for instance. The test failure in that case becomes a distraction, a false positive.It would be ideal to have a configurable timeout for XHR responses or use the same page load timeout to fix the currently hardcoded 20 seconds landmark.

### Specify your

* operating system:
* testcafe version:
* node.js version:

```
$ cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.1 LTS"
$ ./node_modules/.bin/testcafe --version
0.20.3
$ node --version
v8.9.3

```

### PR DESCRIPTION
**Pull request by miherlosev**
Changes:
* add `pageRequestTimeout`, `ajaxRequestTimeout` options
* rewrite `src/test-run/client-messages.js` and `src/test-run/phase.ts` files to TypeScript

Related [PR](https://github.com/DevExpress/testcafe-hammerhead/pull/2470) in hammerhead.

### CODE REVIEW

File: test/functional/fixtures/multiple-windows/test.js [hunk 1]
```diff
@@ -15,287 +15,292 @@ async function assertScreenshotColor (fileName, pixel) {
     expect(assertionHelper.hasPixel(png, pixel, 0, 0)).eql(true);
 }
 
-describe('Multiple windows', () => {
-    describe('Switch to the child window', () => {
-        it('Click on link', () => {
-            return runTests('testcafe-fixtures/switching-to-child/click-on-link.js');
-        });
+if (config.useLocalBrowsers && !config.useHeadlessBrowsers) {
```
Comment (AlexKamaev) [inline test/functional/fixtures/multiple-windows/test.js:18]: There is the `!config.useHeadlessBrowsers` condition. However, if we remove the separate `multiple-windows` task we will have only the `test-functional-local-headless-chrome`. Thus, these tests will be ignored

**Review by AlexKamaev — CHANGES_REQUESTED** [review]

**Comment by miherlosev** [comment]
FPR

**Review by AlexKamaev — APPROVED** [review]

File: src/cli/argument-parser.ts [hunk 2]
```diff
@@ -131,6 +133,8 @@ export default class CLIArgumentParser {
             .option('--selector-timeout <ms>', 'specify the time within which selectors make attempts to obtain a node to be returned')
             .option('--assertion-timeout <ms>', 'specify the time within which assertion should pass')
             .option('--page-load-timeout <ms>', 'specify the time within which TestCafe waits for the `window.load` event to fire on page load before proceeding to the next test action')
+            .option('--page-request-timeout <ms>', 'specify the time within which page request should be completed')
+            .option('--ajax-request-timeout <ms>', 'specify the time within which ajax request (xhr or fetch) should be completed')
```
Comment (VasilyStrelyaev) [inline src/cli/argument-parser.ts:137]:
```
            .option('--page-request-timeout <ms>', 'specify the time within which the request for page HTML should be completed')
             .option('--ajax-request-timeout <ms>', 'specify the time within which AJAX requests (XHR or fetch) should be completed')
```

Comment (VasilyStrelyaev) [reply src/cli/argument-parser.ts:137]: Although... We should wait for Dirk...

Comment (VasilyStrelyaev) [reply src/cli/argument-parser.ts:137]: Checked with Dirk:

```js
            .option('--page-request-timeout <ms>', 'specifies the timeout in milliseconds to complete the request for the page's HTML')
            .option('--ajax-request-timeout <ms>', 'specifies the timeout in milliseconds to complete the AJAX requests (XHR or fetch)')
```

**Review by VasilyStrelyaev — CHANGES_REQUESTED** [review]

**Review by VasilyStrelyaev — APPROVED** [review]

**Review by AndreyBelym — APPROVED** [review]

**Pull request merged**